plugins/dist_table_procedure.cc

```diff
@@ -0,0 +1,581 @@
+    // save if required
+    if(!runProcConf.modelFileUrl.empty()) {
+        filter_ostream stream(runProcConf.modelFileUrl.toString());
+        ML::DB::Store_Writer store(stream);
+        store << distTablesMap;
+    }
+
+    if(!runProcConf.modelFileUrl.empty() && !runProcConf.functionName.empty()) {
+        cerr << "Saving dist tables to " << runProcConf.modelFileUrl.toString() << endl;
+        PolyConfig clsFuncPC;
+        clsFuncPC.type = "distTable.getStats";
+        clsFuncPC.id = runProcConf.functionName;
+        clsFuncPC.params = DistTableFunctionConfig(runProcConf.modelFileUrl);
+
+        cerr << "creating function" << endl;
+        createFunction(server, clsFuncPC, onProgress, true);
+        cerr << "done creating function" << endl;
+    }
+
+    return RunOutput();
+}
+
+
+
+
+
+/*****************************************************************************/
+/* DIST TABLE FUNCTION                                                      */
+/*****************************************************************************/
+DEFINE_STRUCTURE_DESCRIPTION(DistTableFunctionConfig);
+
+DistTableFunctionConfigDescription::
+DistTableFunctionConfigDescription()
+{
+    addField("distTableFileUrl", &DistTableFunctionConfig::modelFileUrl,
+             "URL of the model file (with extension '.dt') to load. "
+             "This file is created by the ![](%%doclink distTable.train procedure).");
+}
+
+DistTableFunction::
+DistTableFunction(MldbServer * owner,
+               PolyConfig config,
+               const std::function<bool (const Json::Value &)> & onProgress)
+    : Function(owner)
+{
+    functionConfig = config.params.convert<DistTableFunctionConfig>();
+
+    // Load saved stats tables
+    filter_istream stream(functionConfig.modelFileUrl.toString());
+    ML::DB::Store_Reader store(stream);
+    store >> distTablesMap;
+}
+
+DistTableFunction::
+~DistTableFunction()
+{
+}
+
+Any
+DistTableFunction::
+getStatus() const
+{
+    Json::Value result;
+    return result;
+}
+
+Any
+DistTableFunction::
+getDetails() const
+{
+    return Any();
+}
+
+ExpressionValue
+DistTableFunction::
+apply(const FunctionApplier & applier,
+      const ExpressionValue & context) const
+{
+    StructValue result;
+
+    ExpressionValue arg = context.getColumn("features");
+
+    if(arg.isRow()) {
+        RowValue rtnRow;
+
+        // TODO should we cache column names as we did in the procedure?
+        auto onAtom = [&] (const ColumnName & columnName,
+                           const ColumnName & prefix,
+                           const CellValue & val,
+                           Date ts)
+            {
+                auto st = distTablesMap.find(columnName);
+                if (st == distTablesMap.end())
+                    return true;
+
+                if (val.empty())
+                    return true;
+
+                const DistTable & distTable = st->second;
+
+                for (int i=0; i < distTable.getNbOutcomes(); ++i) {
```

Comment:
Comme c'est là à chaque itération il fait un appel à `distTable.getNbOutcomes()`. S'il y a beaucoup d'appels et que le résultat est constant, extrait le dans une variable et utilise la variable dans ton for.